recon/src/protocol.rs

```diff
@@ -142,275 +139,360 @@ where
     }
 }
 
-// Protocol manages the state machine of the protocol, delegating to a role implementation.
-struct Protocol<R: Role> {
-    role: R,
-    remote_done: bool,
+type ToWriterSender<M> = mpsc::Sender<ToWriter<M>>;
+
+enum ToWriter<M> {
+    // SyncId learned from the remote.
+    SyncId(String),
+    // Send all messages from the stream to the remote.
+    SendAll(BoxStream<'static, Result<M>>),
+    // Send the messages to the remote as long as its below the work-in-progress limit.
+    // Otherwise messages are queued in LIFO order.
+    SendWIPLimited(Vec<M>),
+    // Mark that in-progress work has completed.
+    WIPCompleted(usize),
+    // Send the finish message as the final message to the remote.
+    // Sender must ensure all in progress work has been completed before sending Finish.
+    Finish,
+}
+
+// Run the recon protocol conversation to completion.
+#[instrument(skip_all)]
+async fn protocol<R, S, E>(
+    sync_id: Option<String>,
+    mut role: R,
+    stream: S,
     metrics: Metrics,
+) -> Result<()>
+where
+    R: Role,
+    R::Out: std::fmt::Debug + Send + 'static,
+    R::In: std::fmt::Debug,
+    MessageLabels: for<'a> From<&'a R::Out>,
+    MessageLabels: for<'a> From<&'a R::In>,
+    S: Stream<Item = Result<ReconMessage<R::In>, E>>
+        + Sink<ReconMessage<R::Out>, Error = E>
+        + Send
+        + 'static,
+    E: std::error::Error + Send + Sync + 'static,
+{
+    let start = Instant::now();
+    let (sink, stream) = stream.split();
+    let (to_writer_tx, to_writer_rx) = mpsc::channel(1000);
+
+    let init = role.init().await?;
+    let write = write(
+        sync_id.clone(),
+        sink.sink_map_err(anyhow::Error::from),
+        to_writer_rx,
+        init,
+        role.finish(),
+        PENDING_RANGES_LIMIT,
+        metrics.clone(),
+    );
+
+    let read = read(sync_id.clone(), stream, role, to_writer_tx, metrics.clone());
```

Comment:
nit: don't think we need to clone sync_id in this line but it probably doesn't actually happen